### PR TITLE
Bunch of changes with streams + basic i/o redirection tests.

### PR DESCRIPTION
**Pull request by rpaquay**
Lots of changes related to decoupling various stream implementations.

@rginda

### CODE REVIEW

File: lib/axiom/fs/nested_stdio.js [hunk 1]
```diff
@@ -45,29 +45,31 @@ export default NestedStdio;
...
  * @return {void}
  */
 NestedStdio.prototype.close = function(error) {
-  this.stdin.close(error);
-  this.signal.close(error);
-  this.ttyin.close(error);
-  this.stdout.close(error);
-  this.stderr.close(error);
-  this.ttyout.close(error);
+  for(var key in this) {
```
Comment (rginda) [inline lib/axiom/fs/nested_stdio.js]: This pattern is a little too shotgun, I think.  Maybe

    ['signal', ..., 'ttyout'].forEach(function(key) {
     ...
    });

would feel less dangerous.

File: lib/axiom/fs/stream/chrome_port_streams.js [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+  this.completer_ = null;
+  /** @const @private @type {BufferedStreams} */
+  this.streams_ = new BufferedStreams(this);
+  /** @const @type {!ReadableStream} */
+  this.readableStream = this.streams_.readableStream;
+  /** @const @type {!WritableStream} */
+  this.writableStream = this.streams_.writableStream;
+
+  // Bind event handlers
+  this.onConnectingMessage_ = this.onConnectingMessage_.bind(this);
```
Comment (rginda) [inline lib/axiom/fs/stream/chrome_port_streams.js]: TODO: not overwrite the existing properties?

File: lib/axiom/fs/stream/node_web_socket_streams.js [hunk 3]
```diff
@@ -14,70 +14,49 @@
...
-  this.readableStream = new ReadableStreamSource(this);
+  this.readableStream = this.streams_.readableStream;
   /** @const @type {!WritableStream} */
-  this.writableStream = new WritableStreamSource(this);
+  this.writableStream = this.streams_.writableStream;
 
-  this.webSocket.on('message', function (message) {
-    this.onSocketMessage_(message);
-  }.bind(this));
+  this.onSocketMessage_ = this.onSocketMessage_.bind(this);
```
Comment (rginda) [inline lib/axiom/fs/stream/node_web_socket_streams.js]: Same TODO here

File: lib/wash/exe/wash.test.js [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+
+/** @typedef {JsExecuteContext$$module$axiom$fs$js$execute_context} */
+var JsExecuteContext;
+
+var failTest = function(error) {
+  console.error('Test failing: ', error, error.stack);
+  expect(error).toBeUndefined();
+};
+
+/*
```
Comment (rginda) [inline lib/wash/exe/wash.test.js]: Remove unnecessary mock?

File: lib/wash/exe/wash.test.js [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+ * Creates an instance of `Wash` given a file system.
+ *
+ * @param {!JsFileSystem} jsfs
+ * @return {!Promise<!Wash>}
+ */
+var createWash = function(jsfs, stdioSource) {
+  return jsfs.createExecuteContext(
+    new Path('jsfs:exe/wash'), stdioSource.stdio, {})
+    .then(function (/** JsExecuteContext */cx) {
+      var washExe = new Wash(cx);
```
Comment (rginda) [inline lib/wash/exe/wash.test.js]: Add a comment calling out the hack involved here?

**Comment by rginda** [comment]
LGTM

File: lib/axiom/fs/nested_stdio.js (hunk 1 above)
Comment (rpaquay) [inline lib/axiom/fs/nested_stdio.js]: Done.

File: lib/axiom/fs/stream/chrome_port_streams.js (hunk 2 above)
Comment (rpaquay) [inline lib/axiom/fs/stream/chrome_port_streams.js]: Done.

File: lib/axiom/fs/stream/node_web_socket_streams.js (hunk 3 above)
Comment (rpaquay) [inline lib/axiom/fs/stream/node_web_socket_streams.js]: Done.

File: lib/wash/exe/wash.test.js (hunk 4 above)
Comment (rpaquay) [inline lib/wash/exe/wash.test.js]: Done.

File: lib/wash/exe/wash.test.js (hunk 5 above)
Comment (rpaquay) [inline lib/wash/exe/wash.test.js]: Done.

**Pull request merged**